### PR TITLE
Lint for Unit-specialization of parameters

### LINKED ISSUE #5507: avoid specializing for Unit in parameter positions
Can we avoid generating "V" specializations in parameter positions? I acknowledge this may be necessary for reasons which my grasp of the possibilities does not make me aware of.  However, consider the substantial impact this small sounding change has.
```scala
final class Foo[@specialized T1, @specialized T2, @specialized R](f: (T1, T2) => R) { }
```
Right now that generates `10*10*10 = 1000` extra classfiles.  If it skipped Unit for T1 and T2, which are completely useless because Unit erases to BoxedUnit and T1 and T2 are already specialized on AnyRef, then it would generate `9*9*10 = 810` classfiles, a 19% reduction.  Or flipped around, we're eating an almost 25% size gain for nothing.

One can of course enumerate the types (or make one of my fancy new groups in the Specializable companion) but the point of the ticket is that it would be even better if we didn't have to tell the compiler that we don't need 25% more custom code to handle all the incoming Units we intend on processing.

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#5507

### CODE REVIEW

**Comment by som-snytt** [comment]
Forgot to do "draft" PR again.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 1]
```diff
@@ -736,6 +738,8 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
 
         if (m.isConstructor) {
           val specCtor = enterMember(cloneInSpecializedClass(m, x => x))
+          if (settings.unitSpecialization && specCtor.paramLists.exists(_.exists(_.typeSignature =:= definitions.UnitTpe)))
+            reporter.warning(m.pos, "Constructor is specialized for a Unit-valued parameter")
```
Comment (eed3si9n) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: let's offer the user what they should do:

```suggestion
            reporter.warning(m.pos, "fruitless specialization of Unit in a parameter position; list specific types as `@specialized(Int, Double, Boolean)` instead")
```

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: `fruitless` is one of my favorite lint words; I wish I had thought of it. Also `bootless`.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 2]
```diff
@@ -608,15 +609,16 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
...
 
-      val sourceFile = clazz.sourceFile
-      if (sourceFile ne null) {
-        sClass.associatedFile = sourceFile
-        currentRun.symSource(sClass) = sourceFile // needed later on by mixin
+      val sourceFile = clazz.sourceFile.tap { sf =>
+        if (sf ne null) {
+          sClass.associatedFile = sf
+          currentRun.symSource(sClass) = sf // needed later on by mixin
+        }
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: to me, using `tap` here doesn't improve readability

**Review by lrytz — COMMENTED** [review]
👍

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala (hunk 2 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: I was getting tapped out. If I'd remembered to mark this a draft PR, I could have punned on draft beer and tap. I will undo the tap dance.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: The pertinent question was, "Is sourceFile even used?" Now that lint is on, we know the answer. It was only used as an alias in the if block.

My other second thought is that I now prefer `val x = v.tap(init)` to mean assign v to x, but do some side-effecting set-up first.

**Comment by som-snytt** [comment]
Addressed the feedback, improved the draconian check, additional syntax tweaks in the source file just to freshen it up.

**Review by eed3si9n — APPROVED** [review]
LGTM

**Comment by som-snytt** [comment]
The CI output prompted me to add the class name:
```
reporter.warning(m.pos, s"Fruitless specialization of Unit in class parameter position for ${m.enclClass}. Consider using `@specialized(Specializable.Arg)` instead.")
```

**Comment by retronym** [comment]
I'm seeing this warning in the build:

```
sbt:root> partest --pos
[info] Compiling 5 Scala sources to /Users/jz/code/scala/build/quick/classes/reflect ...
[info] Compiling 3 Java sources to /Users/jz/code/scala/build/quick/classes/partest-javaagent ...
[warn] Fruitless specialization of Unit in parameter position. Consider using `@specialized(Specializable.Arg)` instead.
[warn] 1 feature warning; re-run with -feature for details
[warn] two warnings found
[info] Compiling 14 Scala sources to /Users/jz/code/scala/build/quick/classes/reflect ...
[warn] 2 deprecations (since 2.13.0)
[warn] 1 deprecation (since 2.13.2)
[warn] 3 deprecations in total; re-run with -deprecation for details
[warn] three warnings found
[info] Compiling 16 Scala sources to /Users/jz/code/scala/build/quick/classes/compiler ...
[warn] Fruitless specialization of Unit in parameter position. Consider using `@specialized(Specializable.Arg)` instead.
```

I'm guessing `m.pos` is `NoPosition` because this check is not restricting itself to symbols from the current compilation run. (The specialization info transform also runs for infos from the classpath.)

**Comment by som-snytt** [comment]
@retronym does that include https://github.com/scala/scala/pull/8943 ?  The details already elude me...

**Pull request merged**